pkg/kapp/permissions/validator.go

```diff
@@ -0,0 +1,508 @@
+					Verb:      "escalate",
+					Group:     res.GroupVersion().Group,
+					Version:   res.GroupVersion().Version,
+					Resource:  mapping.Resource.Resource,
+				},
+			},
+		}
+
+		retSsar, err := rv.ssarClient.Create(ctx, ssar, v1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+
+		if retSsar == nil {
+			return errors.New("returned SelfSubjectAccessReview is nil...")
+		}
+
+		if retSsar.Status.EvaluationError != "" {
+			return errors.New(retSsar.Status.EvaluationError)
+		}
+
+		if retSsar.Status.Allowed {
+			return nil
+		}
+
+		// Check if user has permissions to even create/update the resource
+		ssar = &authv1.SelfSubjectAccessReview{
+			Spec: authv1.SelfSubjectAccessReviewSpec{
+				ResourceAttributes: &authv1.ResourceAttributes{
+					Namespace: res.Namespace(),
+					Name:      res.Name(),
+					Verb:      verb,
+					Group:     res.GroupVersion().Group,
+					Version:   res.GroupVersion().Version,
+					Resource:  mapping.Resource.Resource,
+				},
+			},
+		}
+
+		retSsar, err = rv.ssarClient.Create(ctx, ssar, v1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+
+		if retSsar == nil {
+			return errors.New("returned SelfSubjectAccessReview is nil...")
+		}
+
+		if retSsar.Status.EvaluationError != "" {
+			return errors.New(retSsar.Status.EvaluationError)
+		}
+
+		if !retSsar.Status.Allowed {
+			return fmt.Errorf("not permitted to %q %s, reason: %s",
+				verb,
+				mapping.Resource.String(),
+				retSsar.Status.Reason)
+		}
+		// If user doesn't have "escalate" permissions then they can
+		// only create (Cluster)Roles that contain permissions they already have.
+		// Loop through all the defined policies and determine
+		// if a user has the appropriate permissions
+		rules := []rbacv1.PolicyRule{}
+		switch res.Kind() {
+		case "Role":
+			role := &rbacv1.Role{}
+			err := res.AsTypedObj(role)
+			if err != nil {
+				return fmt.Errorf("converting resource to typed Role object: %w", err)
+			}
+
+			rules = role.Rules
+		case "ClusterRole":
+			role := &rbacv1.ClusterRole{}
+			err := res.AsTypedObj(role)
+			if err != nil {
+				return fmt.Errorf("converting resource to typed Role object: %w", err)
+			}
+
+			rules = role.Rules
+		default:
+			return fmt.Errorf("unknown role kind %q", res.Kind())
+		}
+
+		errorSet := []error{}
+		for _, rule := range rules {
+			// breakdown the rules into the subset of
+			// rules such that the subrules contain
+			// at most one verb, one group, and one resource
+			// source at: https://github.com/kubernetes/component-helpers/blob/9a5801419916272fc9cec7a7822ed525721b99d3/auth/rbac/validation/policy_comparator.go#L56-L84
+			var subrules []rbacv1.PolicyRule = validation.BreakdownRule(rule)
+			for _, subrule := range subrules {
+				ssar := &authv1.SelfSubjectAccessReview{
+					Spec: authv1.SelfSubjectAccessReviewSpec{
+						ResourceAttributes: &authv1.ResourceAttributes{
+							Namespace: res.Namespace(),
+							Verb:      subrule.Verbs[0],
```

Comment:
I don't think there's any way to handle the set of permissions in a Role without having loops for each of these dimensions - it's totally valid and reasonable to specify more than one verb, group, or resource (or all of them!) in a rule, as well as non-resource URLs, etc. That's why the OLMv0 bits we changed recently compute the hash of the rules and expect the user to have pre-created RBAC with the tool, so you can do a simple O(1) lookup.

Reply:
I'd propose that we use `SelfSubjectRulesReview` here instead. I'm working on a kubectl plugin that might help or inspire: https://github.com/joelanford/kubectl-make_rbac

Reply:
>I don't think there's any way to handle the set of permissions in a Role without having loops for each of these dimensions - it's totally valid and reasonable to specify more than one verb, group, or resource (or all of them!) in a rule, as well as non-resource URLs, etc.

Yep, due to the limitation of the `SelfSubjectAccessReview` requiring a single value reference for verb, group, and resource I used a helper that essentially iterates all these values and returns a slice of all the policies broken into a policy where only a single group, resource, and verb value are provided. 


> I'd propose that we use `SelfSubjectRulesReview` here instead. I'm working on a kubectl plugin that might help or inspire: https://github.com/joelanford/kubectl-make_rbac

This is also a direction I'm open to looking into. I've recently made some changes locally that will allow for an easier transition to whatever underlying implementation we feel is better suited for the use case here. Since this is still an early draft and a WIP, I'm going to focus on finishing up the plumbing before making any changes in the "how" the permissions are validated.

Reply:
Based on the discussion from the community meeting, the maintainers would prefer to get this in with SSAR and later investigate updating to use SSRR